Blaze: Add highlight banner on the product list

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #10124 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds the Blaze banner on the product list screen following the requirements in pecCkj-MD-p2. Changes include:
- Added a user defaults key to keep track of the banner dismissal event. 
  - This key is only reset when logging out - it can be annoying to show the banner repeatedly, so I don't reset it when switching stores.
- Updated `ProductListViewModel` to check if the Blaze banner should be displayed and update its visibility when necessary.
- Updated `ProductsViewController` to show the Blaze banner in the table header when there is no loading error and at least one product. If product fetch fails, the error banner is displayed instead.
  - I'm only checking if there is any product loaded. It's not straightforward to check for published products, as it may require fetching and checking all products. Please feel free to advise any solution you see fit. 
- Updated `BlazeBanner` to add an optional divider and spacer to match the design.
- Updated `BlazeBannerHostingController` to handle actions on the banner. 
  - I'm showing only one button on the dismiss alert because the Cancel button seems redundant. Confirming with the designer in [mBhyoV7OPhFnOugxLy6SMS-fi-87:11708] - will update later if they disagree.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Log in as an admin to a public WPCom site with at least one product to ensure you are eligible to create a Blaze campaign.
- Navigate to the Products tab, notice a banner is displayed to highlight Blaze features.
- Notice in Xcode console: `🔵 Tracked blaze_entry_point_displayed, properties: [AnyHashable("source"): "product_list",...]`
- Tap Try Blaze now, the app should navigate to the Blaze flow. An event should be tracked: `🔵 Tracked blaze_entry_point_tapped, properties: [AnyHashable("source"): "product_list",..]`
- Subsequent `blaze_flow_*` events should has the correct source `product_list`.
- Continue to complete creating a campaign (testing tips pe5sF9-1yI-p2#testing-tips) —> an event should be logged 🔵 Tracked blaze_flow_completed, properties: [AnyHashable("source"): "product_list", AnyHashable("current_step"): "step-5”, …]
- Navigate back, notice the Blaze banner is no longer on the product list, even after switching stores back and forth.
- Log out of the app and log back in to make the banner appear again.
- Tap "x" to dismiss the banner. Notice in Xcode console: `🔵 Tracked blaze_banner_dismissed, properties: [ AnyHashable("source"): "product_list",...]`
- Tap Got it on the presented alert, the Blaze banner should be dismissed and will not be available again unless after switching to a different eligible store.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->
| Banner | When dismissed |
| ----- | ----- |
| <img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/26c1e141-1dfc-454a-9d55-5f49bc0dd2ef" width=320 /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/b163c63c-b9e5-4c08-bc1b-cab1ef82fcd0" width=320 /> |



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Add new user defaults for blaze banner

itsmeichigo committed:
Reset the state for Blaze banner upon logging out and switching stores

itsmeichigo committed:
Update ProductsListViewModel to prepare for showing Blaze banner

itsmeichigo committed:
Merge branch 'feat/10099-new-banner-for-blaze' into feat/10124-blaze-banner-product-list

itsmeichigo committed:
Add Blaze banner as header of product list

itsmeichigo committed:
Update blaze icon without the padding

itsmeichigo committed:
Add optional divider and spacer to the banner

itsmeichigo committed:
Use the same user defaults for Blaze banner

itsmeichigo committed:
Keep the banner description fixed size vertically

itsmeichigo committed:
Add closure to handle Blaze campaign created

itsmeichigo committed:
Handle try Blaze button

itsmeichigo committed:
Handle dismissing the blaze banner

itsmeichigo committed:
Do not clear hasDismissedBlazeBanner when switching store

itsmeichigo committed:
Add analytics for displaying and tapping the banner CTA

itsmeichigo committed:
Track banner dismissed

itsmeichigo committed:
Make updateBlazeBannerVisibility async

itsmeichigo committed:
Update unit tests for ProductListViewModel

itsmeichigo committed:
Update test for SessionManager

itsmeichigo committed:
Only check Blaze if there is at least one product

itsmeichigo committed:
Only create new banner if needed

jaclync commented on WooCommerce/Classes/Blaze/BlazeBanner.swift:
```diff
@@ -3,70 +3,140 @@ import struct Yosemite.Site
 final class BlazeBannerHostingController: UIHostingController<BlazeBanner> {
+    private let site: Site
+    private let containerViewController: UIViewController
+    private let dismissHandler: () -> Void
```
nit: this property doesn't seem to be used since it's passed to the action handling closure in the initializer

itsmeichigo replied:
Updated in fe636163fceae30f89bfba5c9010dc94b5e44a5f.

jaclync commented on WooCommerce/Classes/Blaze/BlazeBanner.swift:
```diff
@@ -3,70 +3,140 @@ import struct Yosemite.Site
 ///
 struct BlazeBanner: View {
+    var showsTopDivider: Bool = false
+    var showsBottomSpacer: Bool = false
```
nit: what about `let` so that it doesn't seem like they're mutable?

itsmeichigo replied:
Updated in d0c30a2934a0b7fd33b553e6b563a746a736b05c.

jaclync commented on WooCommerce/Classes/Blaze/BlazeBanner.swift:
```diff
@@ -3,70 +3,140 @@ import struct Yosemite.Site
+            guard let self else { return }
+            analytics.track(event: .Blaze.blazeEntryPointTapped(source: blazeSource))
+            self.showBlaze(source: blazeSource,
+                           onCampaignCreated: dismissHandler)
```
❓ when testing in JPiOS, creating a campaign doesn't dismiss the Blaze banner on the home screen. if users don't want to see it, I think they can just click the dismiss button easily. there might be some users who want to create another campaign after the first try.

also, "hiding" (the equivalent dismissal action) the banner in JPiOS just applies to the selected store. when switching to another store, the Blaze banner is still shown which I think makes sense - the user might not find values promoting in one store but it can be useful for a different store. this is probably an edge case though.

itsmeichigo replied:
> ❓ when testing in JPiOS, creating a campaign doesn't dismiss the Blaze banner on the home screen. if users don't want to see it, I think they can just click the dismiss button easily. there might be some users who want to create another campaign after the first try.

Good point - I forward this question to Ann in the design P2 post: {pecCkj-MD}-p2#comment-397. I will update the behavior in a subsequent PR if needed after getting a confirmation.

> also, "hiding" (the equivalent dismissal action) the banner in JPiOS just applies to the selected store. when switching to another store, the Blaze banner is still shown which I think makes sense - the user might not find values promoting in one store but it can be useful for a different store. this is probably an edge case though.

I've updated the user defaults to use a dictionary with the site ID as the key in 295a720f19284d28e63dae3eadd0be91f5e0b629 and 26bbf0e8368689d07c4d0b8de9e3231206c5e9f3.

jaclync commented on WooCommerce/Classes/ViewRelated/Products/ProductsListViewModel.swift:
```diff
@@ -151,3 +160,27 @@ class ProductListViewModel {
+    func updateBlazeBannerVisibility() async {
+        let isSiteEligible = await blazeEligibilityChecker.isSiteEligible()
+        guard isSiteEligible else {
+            return
+        }
+        guard userDefaults[.hasDismissedBlazeBanner] == nil else {
+            return
```
❓ when reaching the else cases after the first check (since `updateBlazeBannerVisibility` can be called more than once) like when the site becomes ineligible or the user defaults has been reset for some reason, do we want to set `shouldShowBlazeBanner` to `false` as well?

it might also be safer to have a private function that returns the eligibility async to be set to `shouldShowBlazeBanner` since a bool value is returned for each path.

itsmeichigo replied:
Also updated in 26bbf0e8368689d07c4d0b8de9e3231206c5e9f3.

jaclync commented on WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift:
```diff
@@ -854,6 +862,59 @@ private extension ProductsViewController {
+            .removeDuplicates()
+            .sink { [weak self] shouldShow in
+                guard let self else { return }
+                guard !self.hasErrorLoadingData else {
+                    return // ignores Blaze banner if has error fetching data.
+                }
```
❓ If the error banner should be shown while the Blaze banner is already shown, do we want to hide the Blaze banner? or is it already handled when showing the error banner?

itsmeichigo replied:
The [implementation](https://github.com/woocommerce/woocommerce-ios/blob/fa31827aa7a15e345804e29e2af8ec08437aba84/WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift#L768-L770) of the error banner lets the banner replace the header entirely so it's not necessary to hide the banner. But I agree that this is more of an implementation detail, so I've added the call to hide the Blaze banner in case we change something in implementing the error banner.

jaclync commented on WooCommerce/WooCommerceTests/ViewRelated/Products/ProductListViewModelTests.swift:
```diff
@@ -307,4 +307,76 @@ final class ProductListViewModelTests: XCTestCase {
+        // Given
+        let checker = MockBlazeEligibilityChecker(isSiteEligible: true)
+        let uuid = UUID().uuidString
+        let userDefaults = UserDefaults(suiteName: uuid)!
```
nit: how about `try XCTUnwrap(UserDefaults(suiteName: uuid))`? (same for other test cases)

itsmeichigo replied:
Updated in 53c389fedf38946f491adce5fe6ff78731a1f2c1.

jaclync reviewed (approved):
Works as expected in the PR testing steps ✅  just had some questions from the behavior differences from JPiOS.

itsmeichigo committed:
Remove redundant reference to dismissHandler

itsmeichigo committed:
Add extension for UserDefaults to update get/set Blaze banner visibility

itsmeichigo committed:
Update Blaze banner visibility on the product list with the new extension of userDefaults

itsmeichigo committed:
Keep properties constant and private where possible

itsmeichigo committed:
Hide blaze banner when there's an error loading data

itsmeichigo committed:
Replace forced unwrap with XCTUnwrap

itsmeichigo committed:
Fix unit test failures due to user default type change

itsmeichigo committed:
No need to clear hasDismissedBlazeBanner after logging out

itsmeichigo commented:
Thanks @jaclync for the feedback! I updated the code to save the dismissal of the Blaze banner based on site IDs since it makes more sense to save that information for each site. I figure we don't need to clear this information when logging out, so I also removed that in d89bd3d7fb1813482c61105dab2643e2af9fb856.

This PR is ready for another look 🙇‍♀️

itsmeichigo committed:
Merge branch 'trunk' into feat/10124-blaze-banner-product-list

itsmeichigo commented:
Seems like Jaclyn is AFK, @selanthiraiyan could you help take a look for a confidence check? Thanks!

selanthiraiyan commented on WooCommerce/Classes/Extensions/UserDefaults+Woo.swift:
```diff
@@ -94,3 +97,32 @@ extension UserDefaults {
+    /// - Parameters:
+    ///     - siteID: ID of the site to be checked for the Blaze banner visibility.
+    ///
+    func hasDismissedBlazeBanner(for siteID: Int64) -> Bool {
```
Kindly consider adding separate unit tests for these two methods.

itsmeichigo replied:
Oh I missed this! I will add the unit tests in a separate PR, merging this for now.

selanthiraiyan reviewed (approved):
This works as expected. 🚀 

I just left a minor non-blocking comment.
